File: src/core/election/election.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { CreateElectionDto } from './dto/create-election.dto';
import { FindElectionDto } from './dto/find-election.dto';
import { UpdateElectionDto } from './dto/update-election.dto';
import { Election } from './entities/election.entity';
import { IElectionRepository } from './repositories/election.repository';
import { Message } from '../../shared/interfaces/messages';
import { UserService } from '../user/user.service';

@Injectable()
export class ElectionService {
  public constructor(
    private readonly electionRepository: IElectionRepository,
    private readonly userService: UserService,
  ) {}

  public async create(createElectionDto: CreateElectionDto): Promise<Message> {
    await this.userService.findById(createElectionDto.user);
    await this.electionRepository.create(createElectionDto);
    return { message: 'Eleição criada com sucesso' };
  }

  public async findById(id: string): Promise<Election> {
    const election = await this.electionRepository.findById(id);
    if (!election) throw new NotFoundException('Eleição não encontrada');
    return election;
  }

  public findMany(
    user: string,
    findElectionDto: FindElectionDto,
  ): Promise<Election[]> {
    return this.electionRepository.findMany(user, findElectionDto);
  }

  public async update(
    id: string,
    updateElectionDto: UpdateElectionDto,
  ): Promise<Message> {
    await this.findById(id);
    await this.electionRepository.update(id, updateElectionDto);
    return { message: 'Eleição atualizada com sucesso' };
  }

  public async delete(id: string): Promise<Message> {
    await this.findById(id);
    await this.electionRepository.delete(id);
    return { message: 'Eleição removida com sucesso' };
  }
}

File: src/main/java/dev/joaopdias/vox/core/election/ElectionService.java
package dev.joaopdias.vox.core.election;

import dev.joaopdias.vox.core.election.dto.CreateElectionDto;
import dev.joaopdias.vox.core.election.dto.ElectionResponseDto;
import dev.joaopdias.vox.core.election.dto.UpdateElectionDto;
import dev.joaopdias.vox.core.election.entities.Election;
import dev.joaopdias.vox.core.user.UserService;
import dev.joaopdias.vox.core.user.entities.User;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.UUID;

@Service
public class ElectionService {
    private final ElectionRepository electionRepository;
    private final UserService userService;

    public ElectionService(
            ElectionRepository electionRepository,
            UserService userService
    ) {
        this.electionRepository = electionRepository;
        this.userService = userService;
    }

    public ElectionResponseDto create(UUID userId, CreateElectionDto createElectionDto) {
        User user = this.userService.findById(userId);

        Election election = new Election();

        election.setName(createElectionDto.name());
        election.setUser(user);

        this.electionRepository.save(election);

        return election.toResponseDto();
    }

    public Election findById(UUID id) {
        return this.electionRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Eleição não encontrada"));
    }

    public Page<ElectionResponseDto> findManyByUser(UUID userId, Pageable pageable) {
        this.userService.findById(userId);
        return this.electionRepository.findByUserId(userId, pageable).map(Election::toResponseDto);
    }

    public void update(UUID id, UpdateElectionDto updateElectionDto) {
        Election election = this.findById(id);

        if (updateElectionDto.name() != null) election.setName(updateElectionDto.name());

        this.electionRepository.save(election);
    }

    public void delete(UUID id) {
        Election election = this.findById(id);
        this.electionRepository.delete(election);
    }
}
